File: pycut/translation.py
import asyncio
import sys
from typing import List, Optional

try:
    from googletrans import Translator  # type: ignore[import-not-found]
    _TRANSLATOR_IMPORT_ERROR: Optional[ImportError] = None
except ImportError as exc:  # pragma: no cover - exercised via explicit runtime handling
    Translator = None
    _TRANSLATOR_IMPORT_ERROR = exc

MAX_CONSECUTIVE_FAILURES = 3


class GoogleTranslator:
    """Thin synchronous wrapper around the async py-googletrans client."""

    def __init__(self, translator_cls=None):
        self._translator_cls = translator_cls

    def _get_translator_cls(self):
        translator_cls = self._translator_cls or Translator
        if translator_cls is None:
            raise RuntimeError(
                "Translation requires py-googletrans and compatible httpx "
                "(try: pip install -U py-googletrans 'httpx<0.28'). "
                "Original import error: "
                f"{type(_TRANSLATOR_IMPORT_ERROR).__name__}: {_TRANSLATOR_IMPORT_ERROR}"
            ) from _TRANSLATOR_IMPORT_ERROR
        return translator_cls

    async def _translate_async(
        self,
        translator_cls,
        texts: List[str],
        source_lang: str,
        target_lang: str,
    ) -> List[str]:
        async with translator_cls() as translator:
            translations = await translator.translate(texts, src=source_lang, dest=target_lang)
            if isinstance(translations, list):
                return [str(item.text).strip() for item in translations]
            return [str(translations.text).strip()]

    def translate_bulk(
        self,
        texts: List[str],
        source_lang: str = "zh",
        target_lang: str = "en",
    ) -> List[str]:
        if not texts:
            return []

        translator_cls = self._get_translator_cls()
        originals = list(texts)

        for attempt in range(1, MAX_CONSECUTIVE_FAILURES + 1):
            try:
                translated = asyncio.run(
                    self._translate_async(
                        translator_cls,
                        originals,
                        source_lang=source_lang,
                        target_lang=target_lang,
                    )
                )
            except Exception as exc:
                preview = originals[0][:30] if originals else ""
                print(
                    f"⚠️  Translation attempt {attempt}/{MAX_CONSECUTIVE_FAILURES} failed "
                    f"for '{preview}...': {exc}"
                )
                if attempt >= MAX_CONSECUTIVE_FAILURES:
                    print("❌ Translation failed 3 consecutive times. Exiting.")
                    sys.exit(1)
                continue

            if len(translated) != len(originals):
                preview = originals[0][:30] if originals else ""
                print(
                    f"⚠️  Translation returned {len(translated)} items for {len(originals)} texts; "
                    f"keeping originals for '{preview}...'"
                )
                return originals

            return translated

        return originals  # unreachable

    def translate_text(self, text: str, source_lang: str = "zh", target_lang: str = "en") -> str:
        translated = self.translate_bulk([text], source_lang=source_lang, target_lang=target_lang)
        return translated[0] if translated else text

File: src/translation.ts
export const MAX_CONSECUTIVE_FAILURES = 3;

export interface TranslationBackend {
  translateBatch(texts: readonly string[], opts: { from: string; to: string }): Promise<string[]>;
}

export type BackendFactory = () => Promise<TranslationBackend>;

export interface TranslatorLogger {
  warn(message: string): void;
  error(message: string): void;
}

export interface GoogleTranslatorOptions {
  backendFactory?: BackendFactory;
  maxAttempts?: number;
  terminate?: (code: number) => never;
  logger?: TranslatorLogger;
}

const DEFAULT_LOGGER: TranslatorLogger = {
  warn: (m) => console.warn(m),
  error: (m) => console.error(m),
};

function defaultTerminate(code: number): never {
  process.exit(code);
}

export async function defaultGoogleBackendFactory(): Promise<TranslationBackend> {
  let translate: (text: string, options: { from: string; to: string }) => Promise<{ text: string }>;
  try {
    const mod = (await import("@vitalets/google-translate-api")) as {
      translate: typeof translate;
    };
    translate = mod.translate;
  } catch (err) {
    const detail = err instanceof Error ? `${err.name}: ${err.message}` : String(err);
    throw new Error(
      "Translation requires @vitalets/google-translate-api " +
        "(try: bun add @vitalets/google-translate-api). " +
        `Original import error: ${detail}`,
      err instanceof Error ? { cause: err } : undefined,
    );
  }

  return {
    async translateBatch(texts, { from, to }) {
      const out: string[] = [];
      for (const t of texts) {
        const res = await translate(t, { from, to });
        out.push(String(res.text).trim());
      }
      return out;
    },
  };
}

export class GoogleTranslator {
  private readonly backendFactory: BackendFactory;
  private readonly maxAttempts: number;
  private readonly terminate: (code: number) => never;
  private readonly logger: TranslatorLogger;

  constructor(opts: GoogleTranslatorOptions = {}) {
    this.backendFactory = opts.backendFactory ?? defaultGoogleBackendFactory;
    this.maxAttempts = opts.maxAttempts ?? MAX_CONSECUTIVE_FAILURES;
    this.terminate = opts.terminate ?? defaultTerminate;
    this.logger = opts.logger ?? DEFAULT_LOGGER;
  }

  async translateBulk(
    texts: readonly string[],
    sourceLang = "zh",
    targetLang = "en",
  ): Promise<string[]> {
    if (texts.length === 0) return [];

    const originals = [...texts];
    const backend = await this.backendFactory();
    const preview = (originals[0] ?? "").slice(0, 30);

    for (let attempt = 1; attempt <= this.maxAttempts; attempt++) {
      try {
        const translated = await backend.translateBatch(originals, {
          from: sourceLang,
          to: targetLang,
        });
        if (translated.length !== originals.length) {
          this.logger.warn(
            `⚠️  Translation returned ${translated.length} items for ${originals.length} texts; ` +
              `keeping originals for '${preview}...'`,
          );
          return originals;
        }
        return translated;
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        this.logger.warn(
          `⚠️  Translation attempt ${attempt}/${this.maxAttempts} failed for '${preview}...': ${message}`,
        );
        if (attempt >= this.maxAttempts) {
          this.logger.error("❌ Translation failed 3 consecutive times. Exiting.");
          this.terminate(1);
        }
      }
    }

    return originals; // unreachable
  }

  async translateText(text: string, sourceLang = "zh", targetLang = "en"): Promise<string> {
    const translated = await this.translateBulk([text], sourceLang, targetLang);
    return translated[0] ?? text;
  }
}
